include/openssl/sm2.h

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright 2017 The OpenSSL Project Authors. All Rights Reserved.
+ * Copyright 2017 Ribose Inc. All Rights Reserved.
+ * Ported from Ribose contributions from Botan.
+ *
+ * Licensed under the OpenSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#ifndef HEADER_SM2_H
+# define HEADER_SM2_H
+
+# include <openssl/ec.h>
+
+/* The default user id as specified in GM/T 0009-2012 */
+# define SM2_DEFAULT_USERID "1234567812345678"
+
+int SM2_compute_za(uint8_t *out,
+                   const EVP_MD *digest,
+                   const char *user_id, const EC_KEY *key);
+
+/*
+ * SM2 signatures
+ */
+ECDSA_SIG *SM2_do_sign(const EC_KEY *key,
+                       const EVP_MD *digest,
+                       const char *user_id, const uint8_t *msg, size_t msg_len);
+
+int SM2_do_verify(const EC_KEY *key,
+                  const EVP_MD *digest,
+                  const ECDSA_SIG *signature,
+                  const char *user_id, const uint8_t *msg, size_t msg_len);
+
+/*
+ * SM2 encryption
+ */
+size_t SM2_ciphertext_size(const EC_KEY *key,
+                           const EVP_MD *digest,
+                           size_t msg_len);
+
+int SM2_encrypt(const EC_KEY *key,
+                const EVP_MD *digest,
+                const uint8_t *msg,
+                size_t msg_len,
+                uint8_t *ciphertext_buf, size_t *ciphertext_len);
+
+int SM2_decrypt(const EC_KEY *key,
+                const EVP_MD *digest,
+                const uint8_t *ciphertext,
+                size_t ciphertext_len, uint8_t *ptext_buf, size_t *ptext_len);
+
+int ERR_load_SM2_strings(void);
```

Comment:
It's actually preferable to `#include <openssl/sm2err.h>`...

However, I'd like to encourage you to make this header file as well as sm2err.h internal, i.e. have them placed in `crypti/include/internal`...  which means that all `#include <openssl/sm2.h>` and `#include <openssl/sm2err.h>` will have to be changed to `#include "internal/sm2.h"` and `#include "internal/sm2err.h"`...  and remember to change `crypto/err/openssl.ec` accordingly too.

Reply:
OK. I think before doing that I'll have to figure out how to make PKEY methods work. Right now OpenSSL assumes ECC key == ECDSA and there is no way to override it.